net/multi_listen.go

```diff
@@ -0,0 +1,160 @@
+limitations under the License.
+*/
+
+package net
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"sync"
+)
+
+// connErrPair pairs conn and error returned by accept on actual listeners.
+// It is used for communication between main and listener goroutines.
+type connErrPair struct {
+	conn net.Conn
+	err  error
+}
+
+// multiListener implements net.Listener
+type multiListener struct {
+	addrs     []net.Addr
+	listeners []net.Listener
+
+	wg        sync.WaitGroup
+	mu        sync.Mutex
+	connErrCh chan connErrPair
+	closed    bool
+}
+
+// compile time check to ensure *multiListener implements net.Listener.
+var _ net.Listener = &multiListener{}
+
+// MultiListen returns net.Listener which can listen for and accept
+// TCP connections on multiple addresses.
+func MultiListen(ctx context.Context, addresses []string) (net.Listener, error) {
+	return multiListen(
+		ctx,
+		addresses,
+		net.ResolveTCPAddr,
+		func(ctx context.Context, network, address string) (net.Listener, error) {
+			var lc net.ListenConfig
+			return lc.Listen(ctx, network, address)
+		})
+}
+
+// multiListen consumes stdlib dependencies as argument allowing mocking
+// for unit-tests.
+func multiListen(
+	ctx context.Context,
+	addresses []string,
+	resolveTCPAddrFunc func(network, address string) (*net.TCPAddr, error),
+	listenFunc func(ctx context.Context, network, address string) (net.Listener, error),
+) (net.Listener, error) {
+	ml := &multiListener{
+		connErrCh: make(chan connErrPair),
+	}
+
+	if len(addresses) == 0 {
+		return nil, fmt.Errorf("no address provided to listen on")
+	}
+	for _, address := range addresses {
+		addr, err := resolveTCPAddrFunc("tcp", address)
+		if err != nil {
+			// close all listeners
+			_ = ml.Close()
+			return nil, err
+		}
+
+		var network string
+		host, _, err := net.SplitHostPort(addr.String())
+		if err != nil {
+			// close all listeners
+			_ = ml.Close()
+			return nil, err
+		}
+		switch IPFamilyOf(ParseIPSloppy(host)) {
+		case IPv4:
+			network = "tcp4"
+		case IPv6:
+			network = "tcp6"
+		default:
+			// close all listeners
+			_ = ml.Close()
+			return nil, fmt.Errorf("failed to identify ip family of address '%s", addr.String())
+		}
+
+		l, err := listenFunc(ctx, network, addr.String())
+		if err != nil {
+			// close all listeners
+			_ = ml.Close()
+			return nil, err
+		}
+
+		ml.addrs = append(ml.addrs, addr)
+		ml.listeners = append(ml.listeners, l)
+	}
+
+	for _, l := range ml.listeners {
+		ml.wg.Add(1)
+		// spawn a go routine for every listener to wait for incoming connection requests
+		go func(l net.Listener) {
+			defer ml.wg.Done()
+			for {
+				conn, err := l.Accept()
+				ml.mu.Lock()
+				if ml.closed {
+					ml.mu.Unlock()
+					return
+				}
+				ml.connErrCh <- connErrPair{conn: conn, err: err}
+				ml.mu.Unlock()
+			}
+		}(l)
+	}
+	return ml, nil
+}
+
+// Accept is part of net.Listener interface.
+func (ml *multiListener) Accept() (net.Conn, error) {
+	connErr, ok := <-ml.connErrCh
```

Comment:
I don't think this will work. In particular, if a connection arrives when there is not an `Accept()` pending, and then the caller calls `Close()`, it will deadlock, because the listener thread will be blocked trying to write to `ml.connErrCh`, while holding `ml.mu`, so the `Close()` thread will never be able to get `ml.mu`. 

You generally don't want to write to a blocking channel while holding a mutex. (And of course, you can't do it from outside of the mutex the way the code is currently structured because then you might end up writing to it after it's closed.)

The way we usually handle things like this is to have a channel that is _only_ used as a semaphore to wake up the waiting thread, and do the actual data-passing via fields on `ml`, so you don't have to worry about channel buffering. So the listener thread would do

    ml.accepted = append(ml.accepted, connErrPair{conn: conn, err: err})
    // signal acceptReadyCh, unless it has already been signaled
    select {
            case ml.acceptReadyCh <- struct{}{}:
            default:
    }

(The `select` with the `default` keeps it from blocking if the channel is already "full".)

Then the accept thread would check for an item in `ml.accepted`, and if there isn't one, read off `ml.acceptReadyCh` and try again.

(And then various changes to `Close()` to make it all work...)